examples/control-vector-generator/control-vector-generator.cpp

```diff
@@ -0,0 +1,548 @@
+    std::vector<float> b_tensor = std::vector<float>();
+    
+    // random vector gen/norm
+    std::default_random_engine generator(static_cast<unsigned int>(std::time(0)));
+    std::uniform_real_distribution<float> distribution(0.0, 1.0);
+    for (int i = 0; i < cb_data.n_embd; ++i) {
+        b_tensor.push_back(distribution(generator));
+    }
+    normalize_inplace(b_tensor);
+
+    for (int iter = 0; iter < maxIterations; ++iter) {
+
+        // store the previous one so we can check for convergence
+        std::vector<float> b_prev_tensor = b_tensor;
+
+        // matrix multiplication and renormalize
+        b_tensor = mul_mat(matrix, b_tensor, cb_data.n_embd);
+        normalize_inplace(b_tensor);
+
+        // convergence check
+        float diff = 0.0;
+        for (int i = 0; i < cb_data.n_embd; ++i) {
+            diff += std::pow(b_tensor[i] - b_prev_tensor[i], 2);
+        }
+        if (std::sqrt(diff) < tolerance) {
+            break;
+        }
+    }
+
+    return b_tensor;
+}
+
+// TODO translate to ggml
+static void pca(callback_data & cb_data) {
+    size_t n_threads = 8;
+    int    n_layers  = cb_data.v_diff.size();
+    std::vector<std::thread> threads;
+    cb_data.v_final.reserve(n_layers);
+    auto worker_function = [&](int worker_id) {
+        for (int il = worker_id; il < n_layers; il += n_threads) {
+            float * matrix = square_diff(cb_data, il);
+            std::vector<float> eigenvector = power_iteration(cb_data, matrix);
+            cb_data.v_final[il] = &eigenvector[0];
+            delete[] matrix;
+            printf("Done with layer %d\n", il);
+        }
+    };
+    for (int i = 0; i < n_threads; ++i) {
+        threads.emplace_back(worker_function, i);
+    }
+    for (auto & th : threads) th.join();
+    printf("Done with PCA.");
+}
+
+template <typename T>
+static std::string to_string(const T & val) {
+    std::stringstream ss;
+    ss << val;
+    return ss.str();
+}
+
+static void export_gguf(std::vector<float *> v_final, int n_embd, const std::string fname, const std::string model_hint) {
+    struct gguf_context * ctx = gguf_init_empty();
+
+    const std::string arch = "controlvector";
+    gguf_set_val_str(ctx, "general.architecture", arch.c_str());
+    gguf_set_val_str(ctx, (arch + ".model_hint").c_str(), model_hint.c_str());
+    gguf_set_val_i32(ctx, (arch + ".layer_count").c_str(), v_final.size());
+
+    // TODO customize mem size - I have no idea what this is supposed to be
+    struct ggml_init_params params = {
+        /*.mem_size   =*/ ggml_tensor_overhead() * v_final.size(),
+        /*.mem_buffer =*/ NULL,
+        /*.no_alloc   =*/ false,
+    };
+
+    struct ggml_context * ctx_data = ggml_init(params);
+
+    for (size_t i = 0; i < v_final.size(); ++i) {
+        // TODO this number is probably not right - figure out which layer is which
+        // the python implementation uses a dict to handle this, we don't know if it's 1, 2, 3, 4... or other
```

Comment:
I can confirm that on python, the layer number is +1: https://github.com/vgel/repeng/blob/6d709706f1991d04bed82bbf3ce596b0a2b0b812/repeng/extract.py#L285

Also on cpp, the layer index is -1 when loaded: https://github.com/ggerganov/llama.cpp/blob/9022c33646fbf78da35f40c3f98576cc08c40ddf/common/common.cpp#L2723